[Docs] Tidy up the Modal tutorial (#17367)
PrefectHQ/prefect: docs/v3/tutorials/modal.mdx

---
title: Schedule Prefect deployments to run on Modal
description: Learn how to use Modal as a push work pool to schedule, run, and debug your Prefect flows
---

Prefect is the world's best orchestrator.
Modal is the world's best serverless compute framework.
It only makes sense that they would fit together perfectly. 

In this tutorial, you'll build your first Prefect deployment that runs on Modal every hour, and create a scalable framework to grow your project.

Here's an architecture diagram which shows what you'll build.

![image.png](/v3/img/tutorials/modal/modal%200.png)

You'll build a full serverless infrastructure framework for scheduling and provisioning workflows.
GitHub will act as the source of truth, Prefect will be the orchestrator, and Modal will be the serverless infrastructure.

> You'll only pay for only the CPU time you use. 

Reviewer: ```suggestion
You'll only pay for the Modal CPU time you use. 
```

   export PYTHONPATH = .venv

   .PHONY: uv
   uv:
     pip install --upgrade 'uv>=0.5.6,<0.6'
     uv venv

   setup:
     @if [ ! -d ".venv" ] || ! command -v uv > /dev/null; then \
       echo "UV not installed or .venv does not exist, running uv"; \
       make uv; \
     fi
     @if [ ! -f "uv.lock" ]; then \
       echo "Can't find lockfile. Locking"; \
       uv lock; \
     fi
     uv sync --all-extras --all-groups
     uv pip install --no-deps -e .
   ```

1. Set up and manage your environment with `uv`.

   ```bash
   make setup
   ```

1. Create your Python project.

   ```bash
   uv init --lib --name prefect_modal --python 3.12
   ```

   <Info>
   If `uv` creates a `python-version` that is pinned to a micro version (for example, 3.12.6), remove the micro version (for example, 3.12) to avoid downstream issues.
   </Info>

1. Add your dependencies:

   ```bash
   uv add modal 'prefect[github]'
   ```

1. Add the following to the end of the `pyproject.toml` file created by `uv`:

   ```makefile
   [tool.hatch.build.targets.wheel]
   packages = ["src/prefect_modal"]
   ```

1. Initialize the Prefect environment:

   ```bash
   uv run prefect init
   ```

>    On the first question, select `git` as the source control system.

Reviewer: Is it worth including language to explain that `init` will prompt you to choose your code storage?

Author: https://github.com/PrefectHQ/prefect/pull/17367/commits/c06868885ac1013a810615f42fe021dfb022d455

```

<Info>

> Typically, `uv sync` creates a new virtual environment, but Prefect requires you to use root Python.

Reviewer 2: yeah that sounds good to me!

The following steps show how to tell `uv` to use the root Python version.
</Info>

This configuration tells Prefect how to download the code from GitHub into the Modal pod before running the flow.
This is important: you are telling Modal to load the code from `main`.
This means that if you push new code to main, the next run of a flow in Modal will change to whatever is in main.
To keep this stable, you can set that to a release branch, for example, to only update that release branch when you’re stable and ready.

Now, configure a few more options on the `deployment` to get it ready: `entrypoint`, `schedule`, and `description`.

```yaml
  name: prefect-modal-example

>   description: "Your first flow which runs on Modal"

Reviewer: ```suggestion
  description: "Your first deployment which runs daily at 12:00 on Modal"
```

<Tip>
You can run `make prefect-deploy` locally at any time to keep your deployment up to date.
</Tip>

Add the following command to your `Makefile`.

```makefile Makefile
prefect-deploy: setup
    uv run prefect deploy --prefect-file prefect.yaml --all
```

Before pushing your code, run `make prefect-deploy` once to make sure everything works.
Prefect will suggest changes to `prefect.yaml`, which you should accept.

![image.png](/v3/img/tutorials/modal/modal%206.png)

You let Prefect override your deployment spec, which allowed your CD to run without any interruptions or prompts.
If you go to the **Deployment** page in Prefect Cloud, you'll see your deployment!

![image.png](/v3/img/tutorials/modal/modal%207.png)

There are no runs yet.
The output from `make prefect-deploy` shows the command to run your deployed flow.
However, since the code isn’t in GitHub yet, this will fail.
To fix this, push your code:

```bash
git push
```

Go to the repository in **GitHub**, and then open **Actions** to verify that your CI/CD is running successfully.

![image.png](/v3/img/tutorials/modal/modal%208.png)

You can see the flow run in Prefect Cloud as well.

> The flow will run automatically at 12 PM, but you can run it manually.

Reviewer: ```suggestion
The flow will run automatically at 12 PM, but you can also run it manually.
```
